File: Swift/Sources/Astronomy/Astronomical.swift
//
//  Astronomical.swift
//  Adhan
//
//  Copyright © 2018 Batoul Apps. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

import Foundation

struct Astronomical {

    /* The geometric mean longitude of the sun in degrees. */
    static func meanSolarLongitude(julianCentury T: Double) -> Double {
        /* Equation from Astronomical Algorithms page 163 */
        let term1 = 280.4664567
        let term2 = 36000.76983 * T
        let term3 = 0.0003032 * pow(T, 2)
        let L0 = term1 + term2 + term3
        return L0.unwindAngle()
    }

    /* The geometric mean longitude of the moon in degrees. */
    static func meanLunarLongitude(julianCentury T: Double) -> Double {
        /* Equation from Astronomical Algorithms page 144 */
        let term1 = 218.3165
        let term2 = 481267.8813 * T
        let Lp = term1 + term2
        return Lp.unwindAngle()
    }

    static func ascendingLunarNodeLongitude(julianCentury T: Double) -> Double {
        /* Equation from Astronomical Algorithms page 144 */
        let term1 = 125.04452
        let term2 = 1934.136261 * T
        let term3 = 0.0020708 * pow(T, 2)
        let term4 = pow(T, 3) / 450000
        let Ω = term1 - term2 + term3 + term4
        return Ω.unwindAngle()
    }

    /* The mean anomaly of the sun. */
    static func meanSolarAnomaly(julianCentury T: Double) -> Double {
        /* Equation from Astronomical Algorithms page 163 */
        let term1 = 357.52911
        let term2 = 35999.05029 * T
        let term3 = 0.0001537 * pow(T, 2)
        let M = term1 + term2 - term3
        return M.unwindAngle()
    }

    /* The Sun's equation of the center in degrees. */
    static func solarEquationOfTheCenter(julianCentury T: Double, meanAnomaly M: Double) -> Double {
        /* Equation from Astronomical Algorithms page 164 */
        let Mrad = M.degreesToRadians()
        let term1 = (1.914602 - (0.004817 * T) - (0.000014 * pow(T, 2))) * sin(Mrad)
        let term2 = (0.019993 - (0.000101 * T)) * sin(2 * Mrad)
        let term3 = 0.000289 * sin(3 * Mrad)
        return term1 + term2 + term3
    }

    /* The apparent longitude of the Sun, referred to the
     true equinox of the date. */
    static func apparentSolarLongitude(julianCentury T: Double, meanLongitude L0: Double) -> Double {
        /* Equation from Astronomical Algorithms page 164 */
        let longitude = L0 + Astronomical.solarEquationOfTheCenter(julianCentury: T, meanAnomaly: Astronomical.meanSolarAnomaly(julianCentury: T))
        let Ω = 125.04 - (1934.136 * T)
        let λ = longitude - 0.00569 - (0.00478 * sin(Ω.degreesToRadians()))
        return λ.unwindAngle()
    }

    /* The mean obliquity of the ecliptic, formula
     adopted by the International Astronomical Union.
     Represented in degrees. */
    static func meanObliquityOfTheEcliptic(julianCentury T: Double) -> Double {
        /* Equation from Astronomical Algorithms page 147 */
        let term1 = 23.439291
        let term2 = 0.013004167 * T
        let term3 = 0.0000001639 * pow(T, 2)
        let term4 = 0.0000005036 * pow(T, 3)
        return term1 - term2 - term3 + term4
    }

    /* The mean obliquity of the ecliptic, corrected for
     calculating the apparent position of the sun, in degrees. */
    static func apparentObliquityOfTheEcliptic(julianCentury T: Double, meanObliquityOfTheEcliptic ε0: Double) -> Double {
        /* Equation from Astronomical Algorithms page 165 */
        let O: Double = 125.04 - (1934.136 * T)
        return ε0 + (0.00256 * cos(O.degreesToRadians()))
    }

    /* Mean sidereal time, the hour angle of the vernal equinox, in degrees. */
    static func meanSiderealTime(julianCentury T: Double) -> Double {
        /* Equation from Astronomical Algorithms page 165 */
        let JD = (T * 36525) + 2451545.0
        let term1 = 280.46061837
        let term2 = 360.98564736629 * (JD - 2451545)
        let term3 = 0.000387933 * pow(T, 2)
        let term4 = pow(T, 3) / 38710000
        let θ = term1 + term2 + term3 - term4
        return θ.unwindAngle()
    }

    static func nutationInLongitude(solarLongitude L0: Double, lunarLongitude Lp: Double, ascendingNode Ω: Double) -> Double {
        /* Equation from Astronomical Algorithms page 144 */
        let term1 = (-17.2/3600) * sin(Ω.degreesToRadians())
        let term2 =  (1.32/3600) * sin(2 * L0.degreesToRadians())
        let term3 =  (0.23/3600) * sin(2 * Lp.degreesToRadians())
        let term4 =  (0.21/3600) * sin(2 * Ω.degreesToRadians())
        return term1 - term2 - term3 + term4
    }

    static func nutationInObliquity(solarLongitude L0: Double, lunarLongitude Lp: Double, ascendingNode Ω: Double) -> Double {
        /* Equation from Astronomical Algorithms page 144 */
        let term1 =  (9.2/3600) * cos(Ω.degreesToRadians())
        let term2 = (0.57/3600) * cos(2 * L0.degreesToRadians())
        let term3 = (0.10/3600) * cos(2 * Lp.degreesToRadians())
        let term4 = (0.09/3600) * cos(2 * Ω.degreesToRadians())
        return term1 + term2 + term3 - term4
    }

    static func altitudeOfCelestialBody(observerLatitude φ: Double, declination δ: Double, localHourAngle H: Double) -> Double {
        /* Equation from Astronomical Algorithms page 93 */
        let term1 = sin(φ.degreesToRadians()) * sin(δ.degreesToRadians())
        let term2 = cos(φ.degreesToRadians()) * cos(δ.degreesToRadians()) * cos(H.degreesToRadians())
        return asin(term1 + term2).radiansToDegrees()
    }

    static func approximateTransit(longitude L: Double, siderealTime Θ0: Double, rightAscension α2: Double) -> Double {
        /* Equation from page Astronomical Algorithms 102 */
        let Lw = L * -1
        return ((α2 + Lw - Θ0) / 360).normalizeWithBound(max: 1)
    }

    /* The time at which the sun is at its highest point in the sky (in universal time) */
    static func correctedTransit(approximateTransit m0: Double, longitude L: Double, siderealTime Θ0: Double,
                                 rightAscension α2: Double, previousRightAscension α1: Double, nextRightAscension α3: Double) -> Double {
        /* Equation from page Astronomical Algorithms 102 */
        let Lw = L * -1
        let θ = (Θ0 + (360.985647 * m0)).unwindAngle()
        let α = Astronomical.interpolateAngles(value: α2, previousValue: α1, nextValue: α3, factor: m0).unwindAngle()
        let H = (θ - Lw - α).closestAngle()
        let Δm = H / -360
        return (m0 + Δm) * 24
    }

    static func correctedHourAngle(approximateTransit m0: Double, angle h0: Double, coordinates: Coordinates, afterTransit: Bool, siderealTime Θ0: Double,
                                   rightAscension α2: Double, previousRightAscension α1: Double, nextRightAscension α3: Double,
                                   declination δ2: Double, previousDeclination δ1: Double, nextDeclination δ3: Double) -> Double {
        /* Equation from page Astronomical Algorithms 102 */
        let Lw = coordinates.longitude * -1
        let term1 = sin(h0.degreesToRadians()) - (sin(coordinates.latitude.degreesToRadians()) * sin(δ2.degreesToRadians()))
        let term2 = cos(coordinates.latitude.degreesToRadians()) * cos(δ2.degreesToRadians())
        let H0 = acos(term1 / term2).radiansToDegrees()
        let m = afterTransit ? m0 + (H0 / 360) : m0 - (H0 / 360)
        let θ = (Θ0 + (360.985647 * m)).unwindAngle()
        let α = Astronomical.interpolateAngles(value: α2, previousValue: α1, nextValue: α3, factor: m).unwindAngle()
        let δ = Astronomical.interpolate(value: δ2, previousValue: δ1, nextValue: δ3, factor: m)
        let H = (θ - Lw - α)
        let h = Astronomical.altitudeOfCelestialBody(observerLatitude: coordinates.latitude, declination: δ, localHourAngle: H)
        let term3 = h - h0
        let term4 = 360 * cos(δ.degreesToRadians()) * cos(coordinates.latitude.degreesToRadians()) * sin(H.degreesToRadians())
        let Δm = term3 / term4
        return (m + Δm) * 24
    }

    /* Interpolation of a value given equidistant
     previous and next values and a factor
     equal to the fraction of the interpolated
     point's time over the time between values. */
    static func interpolate(value y2: Double, previousValue y1: Double, nextValue y3: Double, factor n: Double) -> Double {
        /* Equation from Astronomical Algorithms page 24 */
        let a = y2 - y1
        let b = y3 - y2
        let c = b - a
        return y2 + ((n/2) * (a + b + (n * c)))
    }

    /* Interpolation of three angles, accounting for
     angle unwinding. */
    static func interpolateAngles(value y2: Double, previousValue y1: Double, nextValue y3: Double, factor n: Double) -> Double {
        /* Equation from Astronomical Algorithms page 24 */
        let a = (y2 - y1).unwindAngle()
        let b = (y3 - y2).unwindAngle()
        let c = b - a
        return y2 + ((n/2) * (a + b + (n * c)))
    }

    /* The Julian Day for a given Gregorian date. */
    static func julianDay(year: Int, month: Int, day: Int, hours: Double = 0) -> Double {

        /* Equation from Astronomical Algorithms page 60 */

        // NOTE: Casting to Int is done intentionally for the purpose of decimal truncation

        let Y: Int = month > 2 ? year : year - 1
        let M: Int = month > 2 ? month : month + 12
        let D: Double = Double(day) + (hours / 24)

        let A: Int = Y/100
        let B: Int = 2 - A + (A/4)

        let i0: Int = Int(365.25 * (Double(Y) + 4716))
        let i1: Int = Int(30.6001 * (Double(M) + 1))
        return Double(i0) + Double(i1) + D + Double(B) - 1524.5
    }

    /* Julian century from the epoch. */
    static func julianCentury(julianDay JD: Double) -> Double {
        /* Equation from Astronomical Algorithms page 163 */
        return (JD - 2451545.0) / 36525
    }

    /* Whether or not a year is a leap year (has 366 days). */
    static func isLeap(year: Int) -> Bool {
        if year % 4 != 0 {
            return false
        }

        if year % 100 == 0 && year % 400 != 0 {
            return false
        }

        return true
    }

    static func seasonAdjustedMorningTwilight(latitude: Double, day: Int, year: Int, sunrise: Date) -> Date {
        let a: Double = 75 + ((28.65 / 55.0) * fabs(latitude))
        let b: Double = 75 + ((19.44 / 55.0) * fabs(latitude))
        let c: Double = 75 + ((32.74 / 55.0) * fabs(latitude))
        let d: Double = 75 + ((48.10 / 55.0) * fabs(latitude))

        let adjustment: Double = {
            let dyy = Double(Astronomical.daysSinceSolstice(dayOfYear: day, year: year, latitude: latitude))
            if ( dyy < 91) {
                return a + ( b - a ) / 91.0 * dyy
            } else if ( dyy < 137) {
                return b + ( c - b ) / 46.0 * ( dyy - 91 )
            } else if ( dyy < 183 ) {
                return c + ( d - c ) / 46.0 * ( dyy - 137 )
            } else if ( dyy < 229 ) {
                return d + ( c - d ) / 46.0 * ( dyy - 183 )
            } else if ( dyy < 275 ) {
                return c + ( b - c ) / 46.0 * ( dyy - 229 )
            } else {
                return b + ( a - b ) / 91.0 * ( dyy - 275 )
            }
        }()

        return sunrise.addingTimeInterval(round(adjustment * -60.0))
    }

    static func seasonAdjustedEveningTwilight(latitude: Double, day: Int, year: Int, sunset: Date) -> Date {
        let a: Double = 75 + ((25.60 / 55.0) * fabs(latitude))
        let b: Double = 75 + ((2.050 / 55.0) * fabs(latitude))
        let c: Double = 75 - ((9.210 / 55.0) * fabs(latitude))
        let d: Double = 75 + ((6.140 / 55.0) * fabs(latitude))

        let adjustment: Double = {
            let dyy = Double(Astronomical.daysSinceSolstice(dayOfYear: day, year: year, latitude: latitude))
            if ( dyy < 91) {
                return a + ( b - a ) / 91.0 * dyy
            } else if ( dyy < 137) {
                return b + ( c - b ) / 46.0 * ( dyy - 91 )
            } else if ( dyy < 183 ) {
                return c + ( d - c ) / 46.0 * ( dyy - 137 )
            } else if ( dyy < 229 ) {
                return d + ( c - d ) / 46.0 * ( dyy - 183 )
            } else if ( dyy < 275 ) {
                return c + ( b - c ) / 46.0 * ( dyy - 229 )
            } else {
                return b + ( a - b ) / 91.0 * ( dyy - 275 )
            }
        }()

        return sunset.addingTimeInterval(round(adjustment * 60.0))
    }

    static func daysSinceSolstice(dayOfYear: Int, year: Int, latitude: Double) -> Int {
        var daysSinceSolstice = 0
        let northernOffset = 10
        let southernOffset = Astronomical.isLeap(year: year) ? 173 : 172
        let daysInYear = Astronomical.isLeap(year: year) ? 366 : 365

        if (latitude >= 0) {
            daysSinceSolstice = dayOfYear + northernOffset
            if (daysSinceSolstice >= daysInYear) {
                daysSinceSolstice = daysSinceSolstice - daysInYear
            }
        } else {
            daysSinceSolstice = dayOfYear - southernOffset
            if (daysSinceSolstice < 0) {
                daysSinceSolstice = daysSinceSolstice + daysInYear
            }
        }

        return daysSinceSolstice
    }
}

File: src/Astronomical.js
/* eslint-disable max-params, max-lines */
import { degreesToRadians, radiansToDegrees, unwindAngle, normalizeToScale, quadrantShiftAngle } from './MathUtils';
import { dateByAddingSeconds } from './DateUtils';

const Astronomical = {

    /* The geometric mean longitude of the sun in degrees. */
    meanSolarLongitude(julianCentury) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 163 */
        const term1 = 280.4664567;
        const term2 = 36000.76983 * T;
        const term3 = 0.0003032 * Math.pow(T, 2);
        const L0 = term1 + term2 + term3;
        return unwindAngle(L0);
    },

    /* The geometric mean longitude of the moon in degrees. */
    meanLunarLongitude(julianCentury) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 144 */
        const term1 = 218.3165;
        const term2 = 481267.8813 * T;
        const Lp = term1 + term2;
        return unwindAngle(Lp);
    },

    ascendingLunarNodeLongitude(julianCentury) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 144 */
        const term1 = 125.04452;
        const term2 = 1934.136261 * T;
        const term3 = 0.0020708 * Math.pow(T, 2);
        const term4 = Math.pow(T, 3) / 450000;
        const Omega = term1 - term2 + term3 + term4;
        return unwindAngle(Omega);
    },

    /* The mean anomaly of the sun. */
    meanSolarAnomaly(julianCentury) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 163 */
        const term1 = 357.52911;
        const term2 = 35999.05029 * T;
        const term3 = 0.0001537 * Math.pow(T, 2);
        const M = term1 + term2 - term3;
        return unwindAngle(M);
    },

    /* The Sun's equation of the center in degrees. */
    solarEquationOfTheCenter(julianCentury, meanAnomaly) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 164 */
        const Mrad = degreesToRadians(meanAnomaly);
        const term1 = (1.914602 - (0.004817 * T) - (0.000014 * Math.pow(T, 2))) * Math.sin(Mrad);
        const term2 = (0.019993 - (0.000101 * T)) * Math.sin(2 * Mrad);
        const term3 = 0.000289 * Math.sin(3 * Mrad);
        return term1 + term2 + term3;
    },

    /* The apparent longitude of the Sun, referred to the
        true equinox of the date. */
    apparentSolarLongitude(julianCentury, meanLongitude) {
        const T = julianCentury;
        const L0 = meanLongitude;
        /* Equation from Astronomical Algorithms page 164 */
        const longitude = L0 + Astronomical.solarEquationOfTheCenter(T, Astronomical.meanSolarAnomaly(T));
        const Omega = 125.04 - (1934.136 * T);
        const Lambda = longitude - 0.00569 - (0.00478 * Math.sin(degreesToRadians(Omega)));
        return unwindAngle(Lambda);
    },

    /* The mean obliquity of the ecliptic, formula
        adopted by the International Astronomical Union.
        Represented in degrees. */
    meanObliquityOfTheEcliptic(julianCentury) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 147 */
        const term1 = 23.439291;
        const term2 = 0.013004167 * T;
        const term3 = 0.0000001639 * Math.pow(T, 2);
        const term4 = 0.0000005036 * Math.pow(T, 3);
        return term1 - term2 - term3 + term4;
    },

    /* The mean obliquity of the ecliptic, corrected for
        calculating the apparent position of the sun, in degrees. */
    apparentObliquityOfTheEcliptic(julianCentury, meanObliquityOfTheEcliptic) {
        const T = julianCentury;
        const Epsilon0 = meanObliquityOfTheEcliptic;
        /* Equation from Astronomical Algorithms page 165 */
        const O = 125.04 - (1934.136 * T);
        return Epsilon0 + (0.00256 * Math.cos(degreesToRadians(O)));
    },

    /* Mean sidereal time, the hour angle of the vernal equinox, in degrees. */
    meanSiderealTime(julianCentury) {
        const T = julianCentury;
        /* Equation from Astronomical Algorithms page 165 */
        const JD = (T * 36525) + 2451545.0;
        const term1 = 280.46061837;
        const term2 = 360.98564736629 * (JD - 2451545);
        const term3 = 0.000387933 * Math.pow(T, 2);
        const term4 = Math.pow(T, 3) / 38710000;
        const Theta = term1 + term2 + term3 - term4;
        return unwindAngle(Theta)
    },

    nutationInLongitude(julianCentury, solarLongitude, lunarLongitude, ascendingNode) {
        const L0 = solarLongitude;
        const Lp = lunarLongitude;
        const Omega = ascendingNode;
        /* Equation from Astronomical Algorithms page 144 */
        const term1 = (-17.2/3600) * Math.sin(degreesToRadians(Omega));
        const term2 =  (1.32/3600) * Math.sin(2 * degreesToRadians(L0));
        const term3 =  (0.23/3600) * Math.sin(2 * degreesToRadians(Lp));
        const term4 =  (0.21/3600) * Math.sin(2 * degreesToRadians(Omega));
        return term1 - term2 - term3 + term4;
    },

    nutationInObliquity(julianCentury, solarLongitude, lunarLongitude, ascendingNode) {
        const L0 = solarLongitude;
        const Lp = lunarLongitude;
        const Omega = ascendingNode;
        /* Equation from Astronomical Algorithms page 144 */
        const term1 =  (9.2/3600) * Math.cos(degreesToRadians(Omega));
        const term2 = (0.57/3600) * Math.cos(2 * degreesToRadians(L0));
        const term3 = (0.10/3600) * Math.cos(2 * degreesToRadians(Lp));
        const term4 = (0.09/3600) * Math.cos(2 * degreesToRadians(Omega));
        return term1 + term2 + term3 - term4;
    },

    altitudeOfCelestialBody(observerLatitude, declination, localHourAngle) {
        const Phi = observerLatitude;
        const delta = declination;
        const H = localHourAngle;
        /* Equation from Astronomical Algorithms page 93 */
        const term1 = Math.sin(degreesToRadians(Phi)) * Math.sin(degreesToRadians(delta));
        const term2 = Math.cos(degreesToRadians(Phi)) * Math.cos(degreesToRadians(delta)) * Math.cos(degreesToRadians(H));
        return radiansToDegrees(Math.asin(term1 + term2));
    },

    approximateTransit(longitude, siderealTime, rightAscension) {
        const L = longitude;
        const Theta0 = siderealTime;
        const a2 = rightAscension;
        /* Equation from page Astronomical Algorithms 102 */
        const Lw = L * -1;
        return normalizeToScale((a2 + Lw - Theta0) / 360, 1);
    },

    /* The time at which the sun is at its highest point in the sky (in universal time) */
    correctedTransit(approximateTransit, longitude, siderealTime, rightAscension, previousRightAscension, nextRightAscension) {
        const m0 = approximateTransit;
        const L = longitude;
        const Theta0 = siderealTime;
        const a2 = rightAscension;
        const a1 = previousRightAscension;
        const a3 = nextRightAscension;
        /* Equation from page Astronomical Algorithms 102 */
        const Lw = L * -1;
        const Theta = unwindAngle((Theta0 + (360.985647 * m0)));
        const a = unwindAngle(Astronomical.interpolateAngles(a2, a1, a3, m0));
        const H = quadrantShiftAngle(Theta - Lw - a);
        const dm = H / -360;
        return (m0 + dm) * 24;
    },

    correctedHourAngle(approximateTransit, angle, coordinates, afterTransit, siderealTime,
                                    rightAscension, previousRightAscension, nextRightAscension, declination, previousDeclination, nextDeclination) {
        const m0 = approximateTransit;
        const h0 = angle;
        const Theta0 = siderealTime;
        const a2 = rightAscension;
        const a1 = previousRightAscension;
        const a3 = nextRightAscension;
        const d2 = declination;
        const d1 = previousDeclination;
        const d3 = nextDeclination;

        /* Equation from page Astronomical Algorithms 102 */
        const Lw = coordinates.longitude * -1;
        const term1 = Math.sin(degreesToRadians(h0)) - (Math.sin(degreesToRadians(coordinates.latitude)) * Math.sin(degreesToRadians(d2)));
        const term2 = Math.cos(degreesToRadians(coordinates.latitude)) * Math.cos(degreesToRadians(d2));
        const H0 = radiansToDegrees(Math.acos(term1 / term2));
        const m = afterTransit ? m0 + (H0 / 360) : m0 - (H0 / 360);
        const Theta = unwindAngle((Theta0 + (360.985647 * m)));
        const a = unwindAngle(Astronomical.interpolateAngles(a2, a1, a3, m));
        const delta = Astronomical.interpolate(d2, d1, d3, m);
        const H = (Theta - Lw - a);
        const h = Astronomical.altitudeOfCelestialBody(coordinates.latitude, delta, H);
        const term3 = h - h0;
        const term4 = 360 * Math.cos(degreesToRadians(delta)) * Math.cos(degreesToRadians(coordinates.latitude)) * Math.sin(degreesToRadians(H));
        const dm = term3 / term4;
        return (m + dm) * 24;
    },

    /* Interpolation of a value given equidistant
        previous and next values and a factor
        equal to the fraction of the interpolated
        point's time over the time between values. */
    interpolate(y2, y1, y3, n) {
        /* Equation from Astronomical Algorithms page 24 */
        const a = y2 - y1;
        const b = y3 - y2;
        const c = b - a;
        return y2 + ((n/2) * (a + b + (n * c)));
    },

    /* Interpolation of three angles, accounting for
        angle unwinding. */
    interpolateAngles(y2, y1, y3, n) {
        /* Equation from Astronomical Algorithms page 24 */
        const a = unwindAngle(y2 - y1);
        const b = unwindAngle(y3 - y2);
        const c = b - a;
        return y2 + ((n/2) * (a + b + (n * c)));
    },

    /* The Julian Day for the given Gregorian date components. */
    julianDay(year, month, day, hours) {
        /* Equation from Astronomical Algorithms page 60 */
        if (typeof hours === 'undefined') {
            hours = 0;
        }

        const trunc = Math.trunc || function (x) { return x < 0 ? Math.ceil(x) : Math.floor(x); };

        const Y = trunc(month > 2 ? year : year - 1);
        const M = trunc(month > 2 ? month : month + 12);
        const D = day + (hours / 24);

        const A = trunc(Y/100);
        const B = trunc(2 - A + trunc(A/4));

        const i0 = trunc(365.25 * (Y + 4716));
        const i1 = trunc(30.6001 * (M + 1));

        return i0 + i1 + D + B - 1524.5;
    },

    /* Julian century from the epoch. */
    julianCentury(julianDay) {
        /* Equation from Astronomical Algorithms page 163 */
        return (julianDay - 2451545.0) / 36525;
    },

    /* Whether or not a year is a leap year (has 366 days). */
    isLeapYear(year) {
        if (year % 4 !== 0) {
            return false;
        }

        if (year % 100 === 0 && year % 400 !== 0) {
            return false;
        }

        return true;
    },

    seasonAdjustedMorningTwilight(latitude, dayOfYear, year, sunrise) {
        const a = 75 + ((28.65 / 55.0) * Math.abs(latitude));
        const b = 75 + ((19.44 / 55.0) * Math.abs(latitude));
        const c = 75 + ((32.74 / 55.0) * Math.abs(latitude));
        const d = 75 + ((48.10 / 55.0) * Math.abs(latitude));

        const adjustment = (function() {
            const dyy = Astronomical.daysSinceSolstice(dayOfYear, year, latitude);
            if ( dyy < 91) {
                return a + ( b - a ) / 91.0 * dyy;
            } else if ( dyy < 137) {
                return b + ( c - b ) / 46.0 * ( dyy - 91 );
            } else if ( dyy < 183 ) {
                return c + ( d - c ) / 46.0 * ( dyy - 137 );
            } else if ( dyy < 229 ) {
                return d + ( c - d ) / 46.0 * ( dyy - 183 );
            } else if ( dyy < 275 ) {
                return c + ( b - c ) / 46.0 * ( dyy - 229 );
            } else {
                return b + ( a - b ) / 91.0 * ( dyy - 275 );
            }
        })();

        return dateByAddingSeconds(sunrise, Math.round(adjustment * -60.0));
    },

    seasonAdjustedEveningTwilight(latitude, dayOfYear, year, sunset) {
        const a = 75 + ((25.60 / 55.0) * Math.abs(latitude));
        const b = 75 + ((2.050 / 55.0) * Math.abs(latitude));
        const c = 75 - ((9.210 / 55.0) * Math.abs(latitude));
        const d = 75 + ((6.140 / 55.0) * Math.abs(latitude));

        const adjustment = (function() {
            const dyy = Astronomical.daysSinceSolstice(dayOfYear, year, latitude);
            if ( dyy < 91) {
                return a + ( b - a ) / 91.0 * dyy;
            } else if ( dyy < 137) {
                return b + ( c - b ) / 46.0 * ( dyy - 91 );
            } else if ( dyy < 183 ) {
                return c + ( d - c ) / 46.0 * ( dyy - 137 );
            } else if ( dyy < 229 ) {
                return d + ( c - d ) / 46.0 * ( dyy - 183 );
            } else if ( dyy < 275 ) {
                return c + ( b - c ) / 46.0 * ( dyy - 229 );
            } else {
                return b + ( a - b ) / 91.0 * ( dyy - 275 );
            }
        })();

        return dateByAddingSeconds(sunset, Math.round(adjustment * 60.0));
    },

    daysSinceSolstice(dayOfYear, year, latitude) {
        let daysSinceSolstice = 0;
        const northernOffset = 10;
        const southernOffset = Astronomical.isLeapYear(year) ? 173 : 172;
        const daysInYear = Astronomical.isLeapYear(year) ? 366 : 365;

        if (latitude >= 0) {
            daysSinceSolstice = dayOfYear + northernOffset;
            if (daysSinceSolstice >= daysInYear) {
                daysSinceSolstice = daysSinceSolstice - daysInYear;
            }
        } else {
            daysSinceSolstice = dayOfYear - southernOffset;
            if (daysSinceSolstice < 0) {
                daysSinceSolstice = daysSinceSolstice + daysInYear;
            }
        }

        return daysSinceSolstice;
    }
};

export default Astronomical;